internal/plugin/available_plugins.go

```diff
@@ -11,3 +13,143 @@ type AvailablePlugins interface {
 	HTTPPlugins() map[string]http.Plugin
 	TCPPlugins() map[string]tcp.Plugin
 }
+
+// Plugins represent a holding object for a bundle of plugins of different
+// types.
+type Plugins struct {
+	HTTPPluginMap map[string]http.Plugin
+	TCPPluginMap  map[string]tcp.Plugin
+}
+
+// HTTPPlugins returns only the HTTP plugins in the Plugins struct.
+func (plugins *Plugins) HTTPPlugins() map[string]http.Plugin {
+	return plugins.HTTPPluginMap
+}
+
+// TCPPlugins returns only the TCP plugins in the Plugins struct.
+func (plugins *Plugins) TCPPlugins() map[string]tcp.Plugin {
+	return plugins.TCPPluginMap
+}
+
+type ExternalPluginLookupFunc func(pluginDir string) ([]*go_plugin.Plugin, error)
+type InternalPluginLookupFunc func() (AvailablePlugins, error)
+
+// InternalPlugins is used to enumerate internally-available plugins to the clients
+// of this method.
+func InternalPlugins(lookupFunc InternalPluginLookupFunc) (AvailablePlugins, error) {
+	plugins, err := lookupFunc()
+	if err != nil {
+		return nil, err
+	}
+
+	if plugins == nil {
+		plugins = &Plugins{
+			HTTPPluginMap: nil,
+			TCPPluginMap:  nil,
+		}
+	}
+
+	return plugins, nil
+}
+
+//TODO: Test this
+// ExternalPlugins is used to enumerate all externally-available plugins in a sepcified
+// directory to the clients of this method.
+func ExternalPlugins(pluginDir string, lookupFunc ExternalPluginLookupFunc) (AvailablePlugins, error) {
+	rawPlugins, err := lookupFunc(pluginDir)
+	if err != nil {
+		return nil, err
+	}
+
+	plugins := Plugins{
+		HTTPPluginMap: map[string]http.Plugin{},
+		TCPPluginMap:  map[string]tcp.Plugin{},
+	}
+
+	for _, rawPlugin := range rawPlugins {
+		pluginInfoSym, err := (*rawPlugin).Lookup("PluginInfo")
+		if err != nil {
+			//TODO: Decide what to do here
+		}
+
+		pluginInfo, ok := pluginInfoSym.(map[string]string)
+		if !ok {
+			//TODO: Decide what to do here
+		}
+
+		// TODO: Do some sort of check on api version
+		// TODO: What do we do if it fails?
+
+		pluginType := pluginInfo["type"]
+		if pluginType == "" {
+			//TODO: Decide what to do here
+		}
+
+		pluginId := pluginInfo["id"]
+		if pluginId == "" {
+			//TODO: Decide what to do here
+			//TODO: What if the plugin id doesn't fit our charset reqs?
+		}
+
+		switch pluginType {
+		case "http":
+			pluginSym, err := (*rawPlugin).Lookup("GetHTTPPlugin")
+			if err != nil {
+				// TODO: Decide what to do here
+			}
+
+			httpPluginFunc, ok := pluginSym.(func() http.Plugin)
+			if !ok {
+				//TODO: Decide what to do here
+			}
+
+			plugins.HTTPPluginMap[pluginId] = httpPluginFunc()
+		case "tcp":
+			pluginSym, err := (*rawPlugin).Lookup("GetTCPPlugin")
+			if err != nil {
+				// TODO: Decide what to do here
+			}
+
+			tcpPluginFunc, ok := pluginSym.(func() tcp.Plugin)
+			if !ok {
+				//TODO: Decide what to do here
+			}
+
+			plugins.TCPPluginMap[pluginId] = tcpPluginFunc()
+		default:
+			//TODO: Decide what to do here
```

Comment:
if it doesn't match http or tcp, that's also not supported, right? so we should write an error and skip loading it